connectivity/drivers/cellular/TELIT/ME910/TELIT_ME910_CellularStack.cpp

```diff
@@ -0,0 +1,793 @@
+
+                    _at.resp_start("\r\nOK", true);
+                    _at.resp_stop();
+
+                    ret = _at.get_last_error();
+
+                    // Set sec level to "Manage server and client authentication if requested by the remote server"
+                    if (ret == NSAPI_ERROR_OK) {
+                        _tls_sec_level = 2;
+                    }
+                }
+                break;
+
+                default:
+                    tr_error("Unsupported sockopt (%d)", optname);
+                    ret = NSAPI_ERROR_UNSUPPORTED;
+            }
+            _at.unlock();
+        } else {
+            tr_error("No optval!");
+            ret = NSAPI_ERROR_PARAMETER;
+        }
+    } else {
+        tr_warning("Unsupported level (%d)", level);
+        ret = NSAPI_ERROR_UNSUPPORTED;
+    }
+
+    return ret;
+}
+
+#ifdef MBED_CONF_CELLULAR_OFFLOAD_DNS_QUERIES
+
+nsapi_error_t mbed::TELIT_ME910_CellularStack::gethostbyname(const char *host,
+        SocketAddress *address, nsapi_version_t version,
+        const char *interface_name) {
+
+    (void) interface_name;
+    MBED_ASSERT(host);
+    MBED_ASSERT(address);
+
+    _at.lock();
+
+    if(_dns_callback) {
+        _at.unlock();
+        return NSAPI_ERROR_BUSY;
+    }
+
+    if(!address->set_ip_address(host)) {
+        _at.set_at_timeout(1min);
+        _at.set_delimiter(',');
+        _at.cmd_start_stop("#QDNS", "=", "%s", host);
+        _at.resp_start("#QDNS: ");
+        if (!read_qdns(*address, version)) {
+            _at.unlock();
+            return NSAPI_ERROR_DNS_FAILURE;
+        }
+        _at.restore_at_timeout();
+        _at.set_default_delimiter();
+    }
+
+    return _at.unlock_return_error();
+
+}
+
+nsapi_value_or_error_t mbed::TELIT_ME910_CellularStack::gethostbyname_async(
+        const char *host, hostbyname_cb_t callback, nsapi_version_t version,
+        const char *interface_name) {
+    (void) interface_name;
+    MBED_ASSERT(host);
+    MBED_ASSERT(callback);
+
+    _at.lock();
+
+    if(_dns_callback) {
+        _at.unlock();
+        return NSAPI_ERROR_BUSY;
+    }
+
+    _at.at_cmd_discard("#QDNS", "=", "%s", host);
+    if(!_at.get_last_error()) {
+        _dns_callback = callback;
+        _dns_version = version;
+    }
+
+    /* The way it is currently implemented, the ME910 only supports one ongoing DNS request at a time */
+    return _at.unlock_return_error() ? NSAPI_ERROR_DNS_FAILURE : 1;
+
+}
+
+nsapi_error_t mbed::TELIT_ME910_CellularStack::gethostbyname_async_cancel(
+        int id) {
+    _at.lock();
+    _dns_callback = nullptr;
+    _at.unlock();
+    return NSAPI_ERROR_OK;
+}
+
+void mbed::TELIT_ME910_CellularStack::urc_qdns() {
+
+    if(!_dns_callback) {
+        return;
+    }
+    SocketAddress address;
+    if(read_qdns(address, _dns_version)) {
+        _dns_callback(1, &address);
+    } else {
+        _dns_callback(NSAPI_ERROR_DNS_FAILURE, nullptr);
+    }
+    _dns_callback = nullptr;
```

Comment:
Stupid question but why mutation of `_dns_callback` requires a lock in `gethostbyname_async_cancel` and not here ?